File: e521.py
import proj_euler as pe
import timeit
import numpy as np
from bisect import bisect_left
import sys

cap = 10**int(sys.argv[1])
mP = min(5*10**7,int(cap**(2./3)))
primes,mask = pe.primes_and_mask(mP)

NP =  len(primes)

print "NUMP " +str(NP)
print "PCAP " + str(mP) 

f = lambda n : n*(n+1) / 2 

s = f(cap)

def prods(prod,sign, startNum,primes,NP):
  global s,cap
  if prod > cap:
    return
  
  for i in xrange(startNum,NP):
    if primes[i] > cap/prod:
      break  
    s += sign * (prod*primes[i] * f(cap/(prod*primes[i])) - primes[i] * (cap/(prod*primes[i])))
    prods(prod*primes[i], -1*sign,i+1,primes,NP)
  return
  

def pi(n):
  global NP
  pos = bisect_left(primes,n)
  if pos == NP:
    return NP
  return pos+1 if primes[pos] == n else pos


#for i, p in enumerate(primes):
 # s -= p * (f(cap/p) - cap/p)
  #for j in xrange(1,i+2):
   # for c in combinations(primes[:i],j):
    #  a = p
     # for x in c:
      # a *= x
      #s += (-1) ** (j-1) * (a*f(cap/a) - p*(cap/a))
#prods(1,-1,0,primes[:pi(int(cap**.5))],pi(int(cap**.5)))
print s-1
s = f(cap) - 1
d = pi(cap/mP)
#print d
#print mP
start = timeit.default_timer()
prods(1,-1,0,primes[:d],d)
print timeit.default_timer()-start
for i in xrange(d,NP):
  if primes[i] > cap/primes[i]:
    break
#  print primes[i]
  up = pi(cap/primes[i])
  lo  = i
#  print "up " + str(up) + " lo " + str(lo)
  pSum = sum(primes[lo:up])
#  print "PSum " + str(pSum)
#  print (primes[i]*pSum - primes[i]*(up - lo + 1))  
  s -= (primes[i]*pSum - primes[i]*(up - lo)) 
  
print s%10**10

File: e521.c
#include "proj_euler.h"

#define CAP 1000000000000LL
#define PCAP 100000000LL
#define NUMP 5761455LL

/*#define CAP 100
#define PCAP 21
#define NUMP*/ 
//typedef long long int LL;
//typedef unsigned long long int ULL;


LL * primes;
LL * psums;

LL s;
LL mod;

LL f(LL n)
{
  if (n % 2==0) return (((n/2)%mod) * ((n+1) % mod)) % mod; 
  else return ((n%mod) * (((n+1)/2)%mod))%mod;
}

LL prods(LL prod, LL startNum,LL sign,LL NP)
{
  if (prod > CAP) return;
  
  LL i;
  
  for (i = startNum;i < NP;i++)
  {
    if (primes[i] > CAP / prod) return;
    LL m = prod * primes[i];
    if (sign)
    {
      s = (s - ((m%mod)*f(CAP/m) % mod) + (primes[i] * (CAP/m)) % mod) % mod;
      if (s < 0) s += mod;
      prods(m,i+1,0,NP);
    }
    
    else
    {
      s = (s + ((m%mod)*f(CAP/m) % mod) - (primes[i] * (CAP/m)) % mod) % mod;    
      if (s < 0) s += mod;
      prods(m,i+1,1,NP);
    }
  } 
}

LL pi(LL n,LL realNumPrimes)
{
  LL pos = big_bin_search(primes,0,realNumPrimes-1,n);
  return pos;
}

int main()
{
  printf("Hello world\n");
  fprintf(stderr,"Hello world\n");
  
  primes = (LL *)malloc(NUMP*sizeof(LL)); 
  psums = (LL *)malloc(NUMP*sizeof(LL));
  char * mask = (char*)malloc(PCAP);
  memset(mask,1,PCAP);
  memset(primes,0,NUMP*sizeof(LL));
  memset(psums,0,NUMP*sizeof(LL));
  LL i,j,k;
  mod = 1000000000LL;
  k = 0;

  printf("started sieving\n");  
  for (i=2;i<PCAP;i++)
  {
    if (mask[i])
    {
      primes[k] = i;
      if (!k) psums[k] = i;
      else psums[k] = (psums[k-1] + i) % mod;
      k++;
      for (j = i+i;j<PCAP;j+=i)
      {
        mask[j] = 0;
      }
    }
  }
  free(mask);
  printf("%lld nump\n",k);    
  fprintf(stderr,"%lld nump\n",k);  
  s = f(CAP);
  
  LL d = pi(CAP/PCAP,k);
  prods(1,0,1,d);
  printf("S after prod %lld\n",s);
  fprintf(stderr,"S after prod %lld\n",s);
  
  for (i = d; i < k; i++)
  {
    if (primes[i] > CAP/primes[i]) break;
    LL hi = pi(CAP/primes[i],k);
//    printf("%lld %lld\n",hi,primes[i]);
    LL psum = psums[hi-1] - psums[i-1];
    s = (s - primes[i]*(psum + i-hi) % mod) % mod;
    if (s < 0) s += mod;
  }
  
  printf("Sum: %lld \n",s-1);
  fprintf(stderr,"Sum: %lld \n",s-1);  
} 
